lib/compat/wordpress-5.9/block-template-utils.php

```diff
@@ -0,0 +1,165 @@
+				'description' => __(
+					'The Footer template defines a page area that typically contains site credits, social links, or any other combination of blocks.',
+					'gutenberg'
+				),
+				'icon'        => 'footer',
+				'area_tag'    => 'footer',
+			),
+		);
+
+		/**
+		 * Filters the list of allowed template part area values.
+		 *
+		 * @param array $default_areas An array of supported area objects.
+		 */
+		return apply_filters( 'default_wp_template_part_areas', $default_area_definitions );
+	}
+}
+
+if ( ! function_exists( '_flatten_blocks' ) ) {
+	/**
+	 * Returns an array containing the references of
+	 * the passed blocks and their inner blocks.
+	 *
+	 * @param array $blocks array of blocks.
+	 *
+	 * @return array block references to the passed blocks and their inner blocks.
+	 */
+	function _flatten_blocks( &$blocks ) {
+		$all_blocks = array();
+		$queue      = array();
+		foreach ( $blocks as &$block ) {
+			$queue[] = &$block;
+		}
+
+		while ( count( $queue ) > 0 ) {
+			$block = &$queue[0];
+			array_shift( $queue );
+			$all_blocks[] = &$block;
+
+			if ( ! empty( $block['innerBlocks'] ) ) {
+				foreach ( $block['innerBlocks'] as &$inner_block ) {
+					$queue[] = &$inner_block;
+				}
+			}
+		}
+
+		return $all_blocks;
+	}
+}
+
+if ( ! function_exists( '_inject_theme_attribute_in_block_template_content' ) ) {
+	/**
+	 * Parses wp_template content and injects the current theme's
+	 * stylesheet as a theme attribute into each wp_template_part
+	 *
+	 * @param string $template_content serialized wp_template content.
+	 *
+	 * @return string Updated wp_template content.
+	 */
+	function _inject_theme_attribute_in_block_template_content( $template_content ) {
+		$has_updated_content = false;
+		$new_content         = '';
+		$template_blocks     = parse_blocks( $template_content );
+
+		$blocks = _flatten_blocks( $template_blocks );
+		foreach ( $blocks as &$block ) {
+			if (
+				'core/template-part' === $block['blockName'] &&
+				! isset( $block['attrs']['theme'] )
+			) {
+				$block['attrs']['theme'] = wp_get_theme()->get_stylesheet();
+				$has_updated_content     = true;
+			}
+		}
+
+		if ( $has_updated_content ) {
+			foreach ( $template_blocks as &$block ) {
+				$new_content .= serialize_block( $block );
+			}
+
+			return $new_content;
+		}
+
+		return $template_content;
+	}
+}
+
+if ( ! function_exists( 'block_template_part' ) ) {
+	/**
+	 * Print a template-part.
+	 *
+	 * @param string $part The template-part to print. Use "header" or "footer".
+	 *
+	 * @return void
+	 */
+	function block_template_part( $part ) {
+		$template_part = gutenberg_get_block_template( get_stylesheet() . '//' . $part, 'wp_template_part' );
+		if ( ! $template_part || empty( $template_part->content ) ) {
+			return;
+		}
+		echo do_blocks( $template_part->content );
+	}
+}
+
+if ( ! function_exists( 'block_header_area' ) ) {
+	/**
+	 * Print the header template-part.
+	 *
+	 * @return void
+	 */
+	function block_header_area() {
```

Comment:
There is no reference for this `block_header_area` and `block_footer_area`. I guess we just want to expose this wrapper util?

Reply:
yes, these are utils to be used by "universal themes"